src/audio/pipeline.c

```diff
@@ -532,6 +532,33 @@ int pipeline_trigger(struct pipeline *p, struct comp_dev *host, int cmd)
 
 	trace_pipe_with_ids(p, "pipeline_trigger()");
 
+	/* handle pipeline global checks before going into each components */
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+		/* if in xrun, prepare before trigger start needed */
+		if (p->xrun_bytes && p->status == COMP_STATE_PAUSED) {
+			trace_pipe_with_ids(p, "in xrun, prepare it first");
+			ret = pipeline_trigger(p, host, COMP_TRIGGER_PREPARE);
+			if (ret < 0) {
+				trace_pipe_error("prepare error: ret = %d",
+						 ret);
+				return ret;
+			}
+			/* now ready for start, clear xrun_bytes */
+			p->xrun_bytes = 0;
+		}
+		break;
+	case COMP_TRIGGER_STOP:
+		/* if in xrun, surpose pipeline is already stop, ignore it */
+		if (p->xrun_bytes && p->status == COMP_STATE_PAUSED) {
```

Comment:
Better to handle this internally in comp_set_state than add explicit check here.

Reply:
this is the behavior for the whole pipeline, not of any specific component, so we need it here.